lib/srv/desktop/rdp/rdpclient/src/rdpdr/mod.rs

```diff
@@ -164,79 +164,85 @@ impl Client {
-            PacketId::PAKID_CORE_CLIENTID_CONFIRM,
-            ClientAnnounceReply::new(req).encode()?,
-        )?;
-        debug!("sending client announce reply");
         Ok(resp)
     }
 
     fn handle_server_capability(&self, payload: &mut Payload) -> RdpResult<Vec<Vec<u8>>> {
         let req = ServerCoreCapabilityRequest::decode(payload)?;
-        debug!("got {:?}", req);
+        debug!("received RDP {:?}", req);
+
+        let resp = ClientCoreCapabilityResponse::new_response();
+        debug!("sending RDP {:?}", resp);
+        let resp =
+            self.add_headers_and_chunkify(PacketId::PAKID_CORE_CLIENT_CAPABILITY, resp.encode()?)?;
 
-        let resp = self.add_headers_and_chunkify(
-            PacketId::PAKID_CORE_CLIENT_CAPABILITY,
-            ClientCoreCapabilityResponse::new_response().encode()?,
-        )?;
-        debug!("sending client core capability response");
         Ok(resp)
     }
 
     fn handle_client_id_confirm(&mut self, payload: &mut Payload) -> RdpResult<Vec<Vec<u8>>> {
         let req = ServerClientIdConfirm::decode(payload)?;
-        debug!("got ServerClientIdConfirm {:?}", req);
+        debug!("received RDP {:?}", req);
 
         // The smartcard initialization sequence that contains this message happens once at session startup,
         // and once when login succeeds. We only need to announce the smartcard once.
         let resp = if !self.active_device_ids.contains(&SCARD_DEVICE_ID) {
             self.push_active_device_id(SCARD_DEVICE_ID)?;
-            self.add_headers_and_chunkify(
-                PacketId::PAKID_CORE_DEVICELIST_ANNOUNCE,
-                ClientDeviceListAnnounceRequest::new_smartcard(SCARD_DEVICE_ID).encode()?,
-            )?
+            let resp = ClientDeviceListAnnounceRequest::new_smartcard(SCARD_DEVICE_ID);
+            debug!("sending RDP {:?}", resp);
+            self.add_headers_and_chunkify(PacketId::PAKID_CORE_DEVICELIST_ANNOUNCE, resp.encode()?)?
         } else {
-            self.add_headers_and_chunkify(
-                PacketId::PAKID_CORE_DEVICELIST_ANNOUNCE,
-                ClientDeviceListAnnounceRequest::new_empty().encode()?,
-            )?
+            let resp = ClientDeviceListAnnounceRequest::new_empty();
+            debug!("sending RDP {:?}", resp);
+            self.add_headers_and_chunkify(PacketId::PAKID_CORE_DEVICELIST_ANNOUNCE, resp.encode()?)?
         };
-        debug!("replying with: {:?}", resp);
         Ok(resp)
     }
 
     fn handle_device_reply(&self, payload: &mut Payload) -> RdpResult<Vec<Vec<u8>>> {
         let req = ServerDeviceAnnounceResponse::decode(payload)?;
-        debug!("got ServerDeviceAnnounceResponse: {:?}", req);
+        debug!("received RDP: {:?}", req);
 
         if !self.active_device_ids.contains(&req.device_id) {
-            (self.tdp_sd_acknowledge)(SharedDirectoryAcknowledge {
+            let resp = SharedDirectoryAcknowledge {
                 err_code: 1,
```

Comment:
I think that it might be helpful if we introduce enum or constants values for error codes. That way it'll be clear what does that code means.

Reply:
Yeah, for sure. We also need them for `file_type` elsewhere (`0` = file, `1` = directory). I've been cutting corners there because I'd tried it, but there was some issue with getting them to work with the C FFI.

But this seems like a good point to get that sorted out, so I will work on that and add it to this PR. For your reference, they're defined in the RFD [here](https://github.com/gravitational/teleport/pull/12103/files#diff-26b655840ed93e16185b07d4813e0a6208feab9f4846c5d575c0bd224b570f7aR464-R467) for now.